### PR TITLE
refator: sensor's architecture redesign

### PR DESCRIPTION
**Pull request by iximiuz**
The redesign is supposed to make the codebase:
- More (unit-)test-friendly.
- More execution-friendly - all (but top-level) errutil.FailOn()
  were removed from the sensor logic.
- Prepared for Execution (not really a context) introduction.
- Prepared for incremental artifact storing.

Introduced abstractions:
- Sensor - to abstract differences in the standalone and controlled execution flows.
- Monitor (fanotify & ptrace) - makes an act of monitoring a noun
  instead of a verb (easier to pass around and interact with).
- CompositeMonitor - to abstract multiple monitors behind single entity and
  simplify the code of the (controlled and standalone) sensors.

Fixed bugs:
- IPC server startup racing with monitor startup - MonitorDoneEvent
  could have been sent before the EvtClient connection leading to sensor
  becoming stuck forver.
- Missed draining of the events channel in ptrace monitor when collection is done
  - files might have been randomly missed in the report.
- Fanotiry report partial initialization leading to panic when the
  monitored process exits relatively fast.

Discovered but not fixed bugs:
- When a file is deleted during tracing, it won't be included in the
  report (even though it should be). Two tests were written for this,
  but they are being skipped at the moment.

### CODE REVIEW

File: pkg/testutil/sensor.go [hunk 1]
```diff
@@ -27,31 +26,50 @@ import (
...
+	commandFileName   = "commands.json"
+	sensorLogFileName = "sensor.log"
+	eventFileName     = "events.json"
 )
 
 var (
 	errNotStarted error = errors.New("test sensor container hasn't been started yet")
 )
 
+type sensorOpt func(*Sensor)
```
Comment (kcq) [inline pkg/testutil/sensor.go:42]: will need to refactor functional options to be regular options (definitely can wait, will be able to do it later)

File: pkg/app/sensor/ipc/ipc.go [hunk 2]
```diff
@@ -101,6 +101,9 @@ func (s *Server) Run() error {
 		return err
 	}
 
+	s.cmdChannel.WaitForConnection()
```
Comment (iximiuz) [inline pkg/app/sensor/ipc/ipc.go:104]: @kcq this is the (hacky) fix for the sensor startup race.

File: pkg/ipc/channel/channel.go [hunk 3]
```diff
@@ -186,20 +187,39 @@ func (s *Server) Stop() {
 
 type EventServer struct {
 	*Server
+
+	mu    sync.Mutex
 	links []net.Conn
+
+	pending chan struct{}
```
Comment (iximiuz) [inline pkg/ipc/channel/channel.go:194]: @kcq the second part of the fix is in this module.

**Comment by kcq** [comment]
can you also create an issue to track the not fixed discoveries

**Comment by iximiuz** [comment]
Here they go:

- #405 
- #406

**Pull request merged**